File: src/main/java/br/com/rotaractsorocabaleste/rotasales/common/jwt/JWTAuthorizationFilter.java
package br.com.rotaractsorocabaleste.rotasales.common.jwt;

import br.com.rotaractsorocabaleste.rotasales.user.UserService;
import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

import static br.com.rotaractsorocabaleste.rotasales.common.jwt.SecurityConstants.HEADER_STRING;
import static br.com.rotaractsorocabaleste.rotasales.common.jwt.SecurityConstants.SECRET;
import static br.com.rotaractsorocabaleste.rotasales.common.jwt.SecurityConstants.TOKEN_PREFIX;

public class JWTAuthorizationFilter extends BasicAuthenticationFilter {

    private final UserService userService;

    public JWTAuthorizationFilter(final AuthenticationManager authManager, final UserService userService) {
        super(authManager);
        this.userService = userService;
    }

    @Override
    protected void doFilterInternal(final HttpServletRequest req,
                                    final HttpServletResponse res,
                                    final FilterChain chain) throws IOException, ServletException {
        final var header = req.getHeader(HEADER_STRING);

        if (header == null || !header.startsWith(TOKEN_PREFIX)) {
            chain.doFilter(req, res);
            return;
        }

        final var authentication = getAuthentication(req);

        SecurityContextHolder.getContext().setAuthentication(authentication);
        chain.doFilter(req, res);
    }

    private UsernamePasswordAuthenticationToken getAuthentication(final HttpServletRequest request) {
        final var token = request.getHeader(HEADER_STRING);

        if (token != null) {
            final var username = JWT.require(Algorithm.HMAC512(SECRET.getBytes()))
                    .build()
                    .verify(token.replace(TOKEN_PREFIX, ""))
                    .getSubject();

            if (username != null) {
                return new UsernamePasswordAuthenticationToken(username, null, userService.findAuthoritiesByUsername(username));
            }

            return null;
        }

        return null;
    }
}

File: src/main/kotlin/br/com/rotaractsorocabaleste/rotasales/core/filter/JWTAuthorizationFilter.kt
package br.com.rotaractsorocabaleste.rotasales.core.filter

import br.com.rotaractsorocabaleste.rotasales.core.utils.JWTUtils
import io.jsonwebtoken.security.SignatureException
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class JWTAuthorizationFilter(
    authenticationManager: AuthenticationManager
) : BasicAuthenticationFilter(authenticationManager) {

    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, chain: FilterChain) {
        val authorizationHeader = request.getHeader("Authorization")
        if (authorizationHeader?.startsWith("Bearer") == true) {
            try {
                val auth = getAuthentication(authorizationHeader)
                SecurityContextHolder.getContext().authentication = auth
            } catch (e: SignatureException) {
                response.contentType = "application/json"
                response.characterEncoding = "UTF-8"
                response.writer.write("{\"error\": \"Invalid token. Please refresh token.\"}")
            }
        }
        chain.doFilter(request, response)
    }

    private fun getAuthentication(authorizationHeader: String?): UsernamePasswordAuthenticationToken {
        val token = authorizationHeader?.substring(7) ?: ""
        if (JWTUtils.isTokenValid(token)) {
            val username = JWTUtils.getUsername(token)
            return UsernamePasswordAuthenticationToken(username, null, mutableListOf())
        }
        throw UsernameNotFoundException("Auth invalid!")
    }

}